primary/src/proposer.rs

```diff
@@ -227,6 +236,40 @@ impl<PublicKey: VerifyingKey> Proposer<PublicKey> {
 
                 // Advance to the next round.
                 self.round += 1;
+
+                // Implement a control system to ensure that some min amount of work
+                // happen in every round, and extend the rounds if it does not. Here we set a target
+                // number of batches committes per round `target_min_batch_per_round`. We then
+                // observe a weighted average of how many batches we commit per round and compute
+                // a multiplier of the max header delay to try to hit this number of rounds
+                // between 1 and `max_multiplier`.
+
+                // The number of rounds since we last committed.
+                let round_diff = self.round - self.metrics.core_metrics.last_commit_round();
+                // The number of additional batches since we last committed.
+                let batch_diff = self.metrics.core_metrics.current_batches()
+                    - self.metrics.core_metrics.last_commit_batches();
+                if round_diff > 0 {
+                    // Compute the weighted average of batches per round included.
+                    ave_batch_per_round = (1.0 - weighted_ave_latest) * ave_batch_per_round
+                        + (batch_diff as f64) * weighted_ave_latest / (round_diff as f64);
+                }
+                debug!(
+                    "Velocity: batch diff / round diff = {} / {} Ave = {:2.2}",
+                    batch_diff, round_diff, ave_batch_per_round
+                );
+
+                // The control system sets log(multiplier) = log(multiplier) + log(target) - log(weighted_ave)
+                // with multiplier clipped between 1 and max_multiplier
+                multiplier = f64::max(
+                    1.0,
+                    f64::min(
+                        max_multiplier,
+                        multiplier * target_min_batch_per_round / (ave_batch_per_round + 0.1),
+                    ),
+                );
```

Comment:
Ah I would like to catch up with you offline @gdanezis about this and go through together. I would also like to better understand how this reacts with spikes in traffic.